README.md
A SNMP library for Dart developers.

Created from templates made available by Stagehand under a BSD-style
[license](https://github.com/almpazel/dart-snmp/blob/master/LICENSE).

## Usage

This package used to SNMP protocol, in flutter also work.

## Tested
- printers;
- .....
- if you need other printers contact with Email.


A simple usage example:

```dart
import 'package:snmp/snmp.dart';

main() {
  //Change your printer ip

  SnmpPack.ip='192.168.199.232';
  SnmpPack.port=161;

  //Printer Total Page
    var pack = SnmpPack();
    var res=await pack.get(OID('1.3.6.1.4.4.4.1'));
    print('response $res');
  
}
```

## References

[Wikipedia/SNMP](https://en.wikipedia.org/wiki/Simple_Network_Management_Protocol) 


CHANGELOG.md
## 0.0.7
- fix update, BER encode and decode
## 0.0.5
- all oid can be request, and respond PDU, can get value PDU.value;
## 0.0.4
- Initial version, created by AlmPazel

## 0.0.1
- Initial version, created by AlmPazel

example/test.dart

import 'dart:convert';

import 'package:convert/convert.dart';
import 'package:snmp/snmp.dart';

Future<void> main() async {

    var oid=OID('1.3.6.1.2.1.43342.10.2.1.4.1.1');
    print(oid);
    print(OID.decode(oid.encode()));
    print(utf8.decode(hex.decode('7075626C6963')));
    print(PDU(OID('1.3.6.1.2.1.43.10.2.1.4.1.1')));
    var pud=PDU(Byter(hex.decode('304C02010004067075626C6963A03F020251B80201000201003033300F060B2B060102011903020105010500300F060B2B060102011903050101010500300F060B2B060102011903050102010500')));
    print(pud);
    print(hex.encode(pud.value.all));

}


lib/src/ber.dart
import 'package:snmp/src/byter.dart';
///
/// About:->
/// Copyright 2020 Alm.Pazel
/// License-Identifier: MIT
///
///
/// Refrences:->
/// https://en.wikipedia.org/wiki/Simple_Network_Management_Protocol
/// https://tools.ietf.org/html/rfc1228
///

class BER {
  static const int ASN_BOOLEAN = 0x01;
  static const int ASN_INTEGER = 0x02;
  static const int ASN_BIT_STR = 0x03;
  static const int ASN_OCTET_STR = 0x04;
  static const int ASN_NULL = 0x05;
  static const int ASN_OBJECT_ID = 0x06;
  static const int ASN_SEQUENCE = 0x10;
  static const int ASN_SET = 0x11;
  static const int ASN_UNIVERSAL = 0x00;
  static const int ASN_APPLICATION = 0x40;
  static const int ASN_CONTEXT = 0x80;
  static const int ASN_PRIVATE = 0xC0;
  static const int ASN_PRIMITIVE = 0x00;
  static const int ASN_CONSTRUCTOR = 0x20;

  static const int ASN_LONG_LEN = 0x80;
  static const int ASN_EXTENSION_ID = 0x1F;
  static const int ASN_BIT8 = 0x80;

  static const int INTEGER = ASN_UNIVERSAL | 0x02;
  static const int INTEGER32 = ASN_UNIVERSAL | 0x02;
  static const int BITSTRING = ASN_UNIVERSAL | 0x03;
  static const int OCTETSTRING = ASN_UNIVERSAL | 0x04;
  static const int NULL = ASN_UNIVERSAL | 0x05;
  static const int OID = ASN_UNIVERSAL | 0x06;
  static const int SEQUENCE = ASN_CONSTRUCTOR | 0x10;

  static const int IPADDRESS = ASN_APPLICATION | 0x00;
  static const int COUNTER = ASN_APPLICATION | 0x01;
  static const int COUNTER32 = ASN_APPLICATION | 0x01;
  static const int GAUGE = ASN_APPLICATION | 0x02;
  static const int GAUGE32 = ASN_APPLICATION | 0x02;
  static const int TIMETICKS = ASN_APPLICATION | 0x03;
  static const int OPAQUE = ASN_APPLICATION | 0x04;
  static const int COUNTER64 = ASN_APPLICATION | 0x06;

  static const int NOSUCHOBJECT = 0x80;
  static const int NOSUCHINSTANCE = 0x81;
  static const int ENDOFMIBVIEW = 0x82;

  static const int LENMASK = 0x0ff;
  static const int MAX_OID_LENGTH = 127;

  static bool checkSequenceLength = true;
  static bool checkValueLength = true;
  static bool checkFirstSubID012 = true;

  static List<int> encodeLength(int length) {
    var os = <int>[];
    if (length < 0) {
      os.add(0x04 | ASN_LONG_LEN);
      os.add((length >> 24) & 0xFF);
      os.add((length >> 16) & 0xFF);
      os.add((length >> 8) & 0xFF);
      os.add(length & 0xFF);
    } else if (length < 0x80) {
      os.add(length);
    } else if (length <= 0xFF) {
      os.add((0x01 | ASN_LONG_LEN));
      os.add(length);
    } else if (length <= 0xFFFF) {
      /* 0xFF < length <= 0xFFFF */
      os.add(0x02 | ASN_LONG_LEN);
      os.add((length >> 8) & 0xFF);
      os.add(length & 0xFF);
    } else if (length <= 0xFFFFFF) {
      /* 0xFFFF < length <= 0xFFFFFF */
      os.add(0x03 | ASN_LONG_LEN);
      os.add((length >> 16) & 0xFF);
      os.add((length >> 8) & 0xFF);
      os.add(length & 0xFF);
    } else {
      os.add(0x04 | ASN_LONG_LEN);
      os.add((length >> 24) & 0xFF);
      os.add((length >> 16) & 0xFF);
      os.add((length >> 8) & 0xFF);
      os.add(length & 0xFF);
    }
    return os;
  }

  static int getOIDLength(List<int> value) {
    var length = 1;
    if (value.length > 1) {
      // for first 2 subids, one sub-id is saved by special encoding
      length = getSubIDLength((value[0] * 40) + value[1]);
    }
    for (var i = 2; i < value.length; i++) {
      length += getSubIDLength(value[i]);
    }
    return length;
  }

  static int getSubIDLength(int subID) {
    int length;
    var v = subID & 0xFFFFFFFF;
    if (v < 0x80) {
      length = 1;
    } else if (v < 0x4000) {
      length = 2;
    } else if (v < 0x200000) {
      length = 3;
    } else if (v < 0x10000000) {
      length = 4;
    } else {
      length = 5;
    }
    return length;
  }

  static dynamic encodeSubID(int subID) {
    var os = <int>[];
    var subid = (subID & 0xFFFFFFFF);
    if (subid < 127) {
      os.add(subid & 0xFF);
    } else {
      var mask = 0x7F; /* handle subid == 0 case */
      var bits = 0;
      /* testmask *MUST* !!!! be of an unsigned type */
      for (var testmask = 0x7F, testbits = 0; testmask != 0; testmask <<= 7, testbits += 7) {
        if ((subid & testmask) > 0) {
          /* if any bits set */
          mask = testmask;
          bits = testbits;
        }
      }
      /* mask can't be zero here */
      for (; mask != 0x7F; mask >>= 7, bits -= 7) {
        /* fix a mask that got truncated above */
        if (mask == 0x1E00000) {
          mask = 0xFE00000;
        }
        os.add((((subid & mask) >> bits) | ASN_BIT8));
      }
      os.add((subid & mask));
    }
    return os;
  }

  static int decodeLength(Byter byter, [bool isCheckLength = true]) {
    var length = 0;
    var lengthbyte = byter.byte();
    if ((lengthbyte & ASN_LONG_LEN) > 0) {
      lengthbyte &= ~ASN_LONG_LEN; /* turn MSb off */
      if (lengthbyte == 0) throw Exception('Indefinite lengths are not supported');
      if (lengthbyte > 4) throw Exception('Data length > 4 bytes are not supported!');
      for (var i = 0; i < lengthbyte; i++) {
        var l = byter.byte() & 0xFF;
        length |= (l << (8 * ((lengthbyte - 1) - i)));
      }
      if (length < 0) throw Exception('SNMP does not support data lengths > 2^31');
    } else {
      length = lengthbyte & 0xFF;
    }
    if (isCheckLength) checkLength(byter, length);
    return length;
  }

  static void checkLength(Byter byter, int length) {
    if (!checkValueLength) return;
    if ((length < 0) || (length > byter.length)) throw Exception('The encoded length $length exceeds,input ${byter.length}');
  }

  static Byter encode(Byter input,[int type=BER.SEQUENCE]) {
    var byter=Byter();
    byter.add(type);
    byter.addAll(encodeLength(input.length));
    byter.addAll(input.all);
    return byter;
  }

  static Byter decode(Byter input,[int type=BER.SEQUENCE]) {
    var t=input.byte();
    if(t!=type) throw Exception('The encoded type $t not equal type:$type ');
    return input.bytes(decodeLength(input));
  }
}

lib/src/oid.dart
import 'ber.dart';
import 'byter.dart';

///
/// About:->
/// Copyright 2020 Alm.Pazel
/// License-Identifier: MIT
///
///
/// Refrences:->
/// https://en.wikipedia.org/wiki/Simple_Network_Management_Protocol
/// https://tools.ietf.org/html/rfc1228
///
class OID {
  final String value;
  List<int> list;

  static const int HEADER = 0x06;

  OID(this.value) {
    var nv=value;
    nv=nv.replaceAll(',', '.');
    list = nv.split('.').map((e) => int.parse(e.trim())).toList();
  }
  String get id=>list.join('.');

  Byter encode() {
    var os = <int>[];
    os.add(HEADER);
    os.addAll(BER.encodeLength(BER.getOIDLength(list)));
    var encodedLength = list.length;
    var rpos = 0;
    if (list.length < 2) {
      os.add(0);
      encodedLength = 0;
    } else {
      var firstSubID = list[0];
      if (BER.checkFirstSubID012 && (firstSubID < 0 || firstSubID > 2)) {
        throw Exception('Invalid first sub-identifier (must be 0, 1, or 2)');
      }
      os.addAll(BER.encodeSubID(list[1] + (firstSubID * 40)));
      encodedLength -= 2;
      rpos = 2;
    }
    while (encodedLength-- > 0) {
      os.addAll(BER.encodeSubID(list[rpos++]));
    }
    return Byter(os);
  }

  @override
  String toString() => 'OID<${list.length}>{ ${list.join('.')} }';

  static OID decode(Byter byter) {
    int subidentifier;
    int length;
    // get the header
    var header = byter.byte();
    if (header != HEADER) throw Exception('Wrong header $HEADER:$header');
    length = BER.decodeLength(byter);
    var oid = List<int>(length+2);
    if (length == 0) oid[0] = oid[1] = 0;
    var pos = 1;
    /* Handle invalid object identifier encodings of the form 06 00 robustly */
    while (length > 0) {
      subidentifier = 0;
      int b;
      do {
        var next = byter.byte();
        if (next < 0) throw Exception('Unexpected end of input stream');
        b = next & 0xFF;
        subidentifier = (subidentifier << 7) + (b & ~BER.ASN_BIT8);
        length--;
      } while ((length > 0) && ((b & BER.ASN_BIT8) != 0));
      /* last byte has high bit clear */
      oid[pos++] = subidentifier;
    }
    subidentifier = oid[1];
    if (subidentifier == 0x2B) {
      oid[0] = 1;
      oid[1] = 3;
    } else if (subidentifier >= 0 && subidentifier < 80) {
      if (subidentifier < 40) {
        oid[0] = 0;
      } else {
        oid[0] = 1;
        oid[1] = subidentifier - 40;
      }
    } else {
      oid[0] = 2;
      oid[1] = subidentifier - 80;
    }
    var noid=[];
    oid.forEach((element) {
      if(element!=null) noid.add(element);
    });
    return OID(noid.join('.'));
  }


  @override
  bool operator ==(other) {
    if(other is OID) return other.id==id;
    return super == other;
  }

}


lib/src/byter.dart
import 'package:convert/convert.dart';

///
/// About:->
/// Copyright 2020 Alm.Pazel
/// License-Identifier: MIT
///
///
/// Refrences:->
/// https://en.wikipedia.org/wiki/Simple_Network_Management_Protocol
/// https://tools.ietf.org/html/rfc1228
///

class Byter {
  final List<int> _bytes = [];

  List<int> get all => _bytes;

  int get first => _bytes.first;

  void clear() => _bytes.clear();

  @override
  String toString() => 'Byter{${length}}';
  String toHexString() => hex.encode(all);

  Byter([dynamic input]) {
    _bytes.clear();
    if (input is List<int>) _bytes.addAll(input);
    if (input is String) _bytes.addAll(hex.decode(input));
  }

  int get length => _bytes.length;

  bool get isEmpty => _bytes.isEmpty;

  bool get isNotEmpty => _bytes.isNotEmpty;

  Byter bytes([int len = 1]) {
    var r = _bytes.sublist(0, len);
    _bytes.removeRange(0, len);
    return Byter(r);
  }

  int byte() => bytes(1)?.first;

  void add(dynamic b) {
    if(b is int) _bytes.add(b);
    if(b is Byter) addAll(b.all);
  }

  void addAll(List<int> os) => os.forEach(add);

  void eat(dynamic b){
    if(b is int) _bytes.insert(0, b);
    if(b is Byter) eatAll(b.all);
  }

  void eatAll(List<int> os) => os.reversed.forEach(eat);

  Byter clone()  => Byter(all);

  int toInt({int radix=16}) =>int.parse(toHexString(),radix: radix);

}


lib/src/snmp_base.dart
import 'dart:async';
import 'dart:io';

import 'byter.dart';
import 'oid.dart';
import 'pdu.dart';

export 'oid.dart';
export 'ber.dart';
export 'pdu.dart';
export 'byter.dart';
///
/// About:->
/// Copyright 2020 Alm.Pazel
/// License-Identifier: MIT
///
///
/// Refrences:->
/// https://en.wikipedia.org/wiki/Simple_Network_Management_Protocol
/// https://tools.ietf.org/html/rfc1228
///

class SnmpPack {
  RawDatagramSocket socket;
  Map<String,Completer<PDU>> completer={};
  SnmpPack(RawDatagramSocket datagramSocket){
    socket=datagramSocket;
    socket.listen(onSubscribe);
  }

  void onSubscribe(RawSocketEvent event) {
    //print(event.toString());
    try{
      var dg = socket.receive();
      if(dg != null) {
        var pdu=PDU(Byter(dg.data.toList()));
//        print('res:$pdu');
        if(pdu!=null&&completer.containsKey(pdu.id)){
          completer[pdu.id].complete(pdu);
        }
      }
    }catch(e){
      print(event.toString());
    }
  }

  /// @oid example OID('1.3.6.xxxx.x.x')
  /// @to example 0.0.0.0:161
  Future<PDU> get(PDU pdu,String to,{Duration timeOut}) async {
    var uri=to.split(':');
    completer[pdu.id]=Completer<PDU>();
    socket.send(pdu.bytes(), InternetAddress(uri.first), int.parse(uri.last));
    var res=await completer[pdu.id].future.timeout(timeOut??Duration(seconds: 5),onTimeout:(){
      return null;
    });
    completer.remove(pdu.id);
    return res;
  }

  static Future<SnmpPack> bind({InternetAddress address,int port=1614}) async {
    return SnmpPack(await RawDatagramSocket.bind(address??InternetAddress.anyIPv4, port));
  }

  void close() {
    try{
      socket.close();
    // ignore: empty_catches
    }catch(e){
    }
  }

}


lib/src/pdu.dart
import 'dart:convert';

import 'package:convert/convert.dart';
import 'package:snmp/snmp.dart';
import 'package:snmp/src/byter.dart';

import 'ber.dart';

///
/// About:->
/// Copyright 2020 Alm.Pazel
/// License-Identifier: MIT
///
///
/// Refrences:->
/// https://en.wikipedia.org/wiki/Simple_Network_Management_Protocol
/// https://tools.ietf.org/html/rfc1228
///
class PDU {
  static const int HEADER = 0x30;
  static const int GET = (BER.ASN_CONTEXT | BER.ASN_CONSTRUCTOR);
  static const int GETNEXT = (BER.ASN_CONTEXT | BER.ASN_CONSTRUCTOR | 0x1);
  static const int RESPONSE = (BER.ASN_CONTEXT | BER.ASN_CONSTRUCTOR | 0x2);
  static const int SET = (BER.ASN_CONTEXT | BER.ASN_CONSTRUCTOR | 0x3);
  static const int V1TRAP = (BER.ASN_CONTEXT | BER.ASN_CONSTRUCTOR | 0x4);
  static const int GETBULK = (BER.ASN_CONTEXT | BER.ASN_CONSTRUCTOR | 0x5);
  static const int INFORM = (BER.ASN_CONTEXT | BER.ASN_CONSTRUCTOR | 0x6);
  static const int TRAP = (BER.ASN_CONTEXT | BER.ASN_CONSTRUCTOR | 0x7);
  static const int NOTIFICATION = TRAP;
  static const int REPORT = (BER.ASN_CONTEXT | BER.ASN_CONSTRUCTOR | 0x8);

  OID oid;

  int version=0;

  String community ='public';

  Byter raw = Byter();
  Byter value = Byter();

  int requestId = 0000000001;
  int requestStatus = 0;
  int requestIndex = 0;

  String get id => oid.id;

  @override
  String toString() {
    return 'PDU<$requestId,$requestStatus,$requestIndex>{ $oid->$value, community:$community, raw:$raw }';
  }

  String toHexString() => raw.toHexString();

  Map _option = {};

  Byter body = Byter();
  /// input should be OID, Byter , hexStr
  PDU(dynamic input, {Map option}) {
    _option = option ?? {};
    if (input is OID) encode(input);
    if (input is Byter) decode(input);
    if (input is String) decode(Byter(input));
  }

  static int get timeint => DateTime.now().millisecondsSinceEpoch;

  void requestEncode() {}

  void decode(Byter input) {
    raw = input.clone();
    var byter = BER.decode(input);
    var _version = BER.decode(byter, 0x02);
    version=int.parse(_version.toHexString(),radix: 16);
    community = utf8.decode(BER.decode(byter, 0x04).all);
    var type=byter.byte();
    byter.eat(type);
    body = BER.decode(byter, type);
    var req = body.clone();
    var reqId = BER.decode(req, 0x02);
    requestId = int.parse(reqId.toHexString(), radix: 16);
    var reqState = BER.decode(req, 0x02);
    requestStatus = int.parse(reqState.toHexString(), radix: 16);
    var reqIndex = BER.decode(req, 0x02);
    requestStatus = int.parse(reqIndex.toHexString(), radix: 16);
    var v1 = BER.decode(req);
    var v2 = BER.decode(v1);
    oid = OID.decode(v2);
    value = v2;
  }

  void encode(OID oid) {
    this.oid = oid;
    var oids = oid.encode(); //oid
    oids.add(0x05);
    oids.add(0x00);
    var req = BER.encode(BER.encode(oids)); //var-bind
    req.eat(BER.encode(Byter([0]), 0x02)); //request index
    req.eat(BER.encode(Byter([0]), 0x02)); //request status
    requestId = (timeint / 1000).round();
    req.eat(BER.encode(Byter(hex.decode(requestId.toRadixString(16))), 0x02)); //request id
    body = BER.encode(req, GET);
    var _pdu = Byter();
    _pdu.add(BER.encode(Byter([version]), 0x02)); //version info
    _pdu.add(BER.encode(Byter(community.codeUnits), 0x04)); //community
    _pdu.add(body);
    raw = BER.encode(_pdu,HEADER);
  }

  List<int> bytes() => raw.all;

  static PDU fromOid(String s) => PDU(OID(s));
}
